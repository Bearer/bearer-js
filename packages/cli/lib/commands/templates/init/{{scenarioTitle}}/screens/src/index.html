<!DOCTYPE html>
<html dir="ltr" lang="en">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0">
    <title>{{scenarioTitle}} scenario</title>
    <script src="/build/{{componentTagName}}.js"></script>
    <style>
      body {
        padding: 50px;
      }
      hr {
        margin: 20px 0;
      }
    </style>
  </head>
  <body>
    <h1>Setup component</h1>
    <{{componentTagName}}-setup></{{componentTagName}}-setup>
    <hr />
    <h1>Action component</h1>
    <{{componentTagName}}></{{componentTagName}}>
    <hr />
    <h1>Display component</h1>
    <{{componentTagName}}-display></{{componentTagName}}-display>
  </body>
</html>
