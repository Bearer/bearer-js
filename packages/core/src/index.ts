import bearer from '@bearer/js'

import Bearer from './bearer'
import classNames from './classnames'
import * as Debug from './debug'
import EventNames from './event-names'
import * as Requests from './requests'
import { TTranslatorFunc, TPluralizerFunc, scopedPluralize, scopedTranslate } from './i18n/index'
export { TTranslatorFunc, TPluralizerFunc } from './i18n/index'

import * as bearerState from './bearer-state'
export * from './decorators'
export const StateManager = bearerState
export const Events = EventNames
export const requests = Requests
export const debug = Debug
export const classnames = classNames

// Next 2 helpers get rewritten with transpiler

export declare const t: TTranslatorFunc

export declare const p: TPluralizerFunc

export const scopedT = (scope: string) => scopedTranslate(scope)(bearer.i18n)
export const scopedP = (scope: string) => scopedPluralize(scope)(bearer.i18n)

export default Bearer

if (process.env.BUILD !== 'distribution') {
  console.warn(`[BEARER] Running non production Bearer Core lib | version ${Bearer.version}`)
}
