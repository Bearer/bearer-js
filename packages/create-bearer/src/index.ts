#!/usr/bin/env node
import * as fs from 'fs'
import * as path from 'path'
const { run } = require('@bearer/cli/lib/index')

const bearer = fs.readFileSync(path.join(__dirname, '../static/message.txt'), { encoding: 'utf8' })
console.log(bearer)

run(['new', ...process.argv.slice(2)])
  .then()
  .catch(() => {
    process.exit(0)
  })
