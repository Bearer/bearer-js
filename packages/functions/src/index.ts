import { DBClient as CLIENT } from './db-client'

export * from './declaration'
export { SaveState } from './functions/save-state'
export { FetchData } from './functions/fetch'

// tslint:disable-next-line:variable-name
export const DBClient = CLIENT.instance
