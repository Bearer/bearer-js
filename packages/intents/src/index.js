const { sendSuccessMessage, sendErrorMessage } = require('./lambda')

const Intent = {
  getCollection: (callback, { collection }) => {
    if (collection) {
      sendSuccessMessage(callback, collection)
    } else {
      sendErrorMessage(callback, { error: 'Error' })
    }
  },

  getObject: (callback, { object }) => {
    if (object) {
      sendSuccessMessage(callback, object)
    } else {
      sendErrorMessage(callback, { error: 'Error' })
    }
  }
}

const StoreObject = {
  intent(action) {
    return (event, context, callback) =>
      action(
        event.accessToken,
        event.queryStringParameters,
        event.body,
        result => {
          Intent.getObject(callback, result)
        }
      )
  }
}

const GetCollection = {
  intent(action) {
    return (event, context, callback) =>
      action(event.accessToken, event.queryStringParameters, result => {
        Intent.getCollection(callback, result)
      })
  }
}

const GetObject = {
  intent(action) {
    return (event, context, callback) =>
      action(event.accessToken, event.queryStringParameters, result => {
        Intent.getObject(callback, result)
      })
  }
}

module.exports = { Intent, GetCollection, GetObject, StoreObject }
