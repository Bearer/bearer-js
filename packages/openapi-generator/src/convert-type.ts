import ts from 'typescript'
import merge from 'lodash.merge'

function isPrimitive(flags: number): boolean {
  return (
    (flags & ts.TypeFlags.String) !== 0 || (flags & ts.TypeFlags.Number) !== 0 || (flags & ts.TypeFlags.Boolean) !== 0
  )
}

export function convertType(
  typ: ts.Type | ts.UnionOrIntersectionType,
  checker: ts.TypeChecker,
  definition: any = {}
): any {
  const flags = typ.getFlags()
  const typeNode = checker.typeToTypeNode(typ)!
  if (ts.isArrayTypeNode(typeNode)) {
    let elementType = (checker as any).getElementTypeOfArrayType(typ)
    const elementTypeFlags = elementType.getFlags()
    if (!isPrimitive(elementTypeFlags)) {
      elementType = checker.getApparentType(elementType)
    }
    return { type: 'array', items: convertType(elementType, checker) }
  }
  if (flags & ts.TypeFlags.String) {
    return { type: 'string' }
  }
  if (flags & ts.TypeFlags.Number) {
    return { type: 'number' }
  }
  if (flags & ts.TypeFlags.Boolean) {
    return { type: 'boolean' }
  }
  if (flags & ts.TypeFlags.Intersection || flags & ts.TypeFlags.Union) {
    return (typ as ts.IntersectionType).types.reduce((acc, t) => {
      const converted = convertType(t, checker, acc)
      return merge(acc, converted)
    }, {})
  }

  if (flags & ts.TypeFlags.Object) {
    return {
      type: 'object',
      properties: checker
        .getWidenedType(typ)
        .getProperties()
        .reduce((acc: any, s): any => {
          const nestedType = checker.getTypeOfSymbolAtLocation(s, s.valueDeclaration!)
          acc[checker.symbolToString(s)] = convertType(checker.getWidenedType(nestedType), checker)
          return acc
        }, {})
    }
  }
  if (typ.aliasTypeArguments && typ.aliasTypeArguments.length) {
    return typ.aliasTypeArguments.reduce((acc, t) => {
      acc[checker.symbolToString(t.symbol)] = convertType(t, checker)
      return acc
    }, definition)
  }

  if (typ.aliasSymbol) {
    const aliasedType = checker.getTypeOfSymbolAtLocation(typ.aliasSymbol, typ.aliasSymbol.valueDeclaration!)
    return convertType(aliasedType, checker, definition)
  }
  return definition
}
