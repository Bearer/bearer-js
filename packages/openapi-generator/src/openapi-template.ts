type THeader = {
  openapi: string
  info: {
    title: string
  }
  servers: { url: string }[]
  tags: {}[]
  components: {
    securitySchemes: {
      [key: string]: {
        type: string
        in: string
        description: string
        name: string
      }
    }
  }
}

type TParam = {
  name: string
  schema: { type: string; format: string }
  in: string
  description: string
  required?: boolean
}

export function specPath({
  buid,
  functionName,
  requestBody,
  response,
  oauth
}: {
  buid: string
  functionName: string
  requestBody: any
  response: any
  oauth: boolean
}) {
  return {
    [`/${buid}/${functionName}`]: {
      post: {
        parameters: [oauthParam(oauth)].filter(e => e !== undefined),
        summary: functionName,
        requestBody: { content: { 'application/json': { schema: requestBody } } },
        responses: {
          '200': { description: functionName, content: { 'application/json': { schema: response } } },
          '401': {
            description: 'Access forbidden',
            content: {
              'application/json': { schema: { type: 'object', properties: { error: { type: 'string' } } } }
            }
          },
          '403': {
            description: 'Unauthorized',
            content: {
              'application/json': { schema: { type: 'object', properties: { error: { type: 'string' } } } }
            }
          }
        },
        security: [
          {
            'API Key': []
          }
        ]
      }
    }
  }
}

function oauthParam(oauth: boolean): TParam | undefined {
  if (oauth) {
    return {
      name: 'authId',
      schema: { type: 'string', format: 'uuid' },
      in: 'query',
      description: 'User Identifier',
      required: true
    }
  }
  return
}

export function topOfSpec(integrationName: string): THeader {
  return {
    openapi: '3.0.0',
    info: {
      title: integrationName
    },
    servers: [{ url: 'https://int.bearer.sh/api/v4/functions/backend/' }],
    tags: [],
    components: {
      securitySchemes: {
        'API Key': {
          type: 'apiKey',
          in: 'header',
          name: 'Authorization',
          description: 'You can find your **Bearer API Key** [here](https://app.bearer.sh/keys).'
        }
      }
    }
  }
}
