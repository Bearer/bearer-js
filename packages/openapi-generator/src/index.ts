import path from 'path'
import ts from 'typescript'
import { topOfSpec, specPath } from './openapi-template'

import { convertType } from './convert-type'

export const FUNCTION_ACTION = 'action'
type TOutput = { requestBody: any; response: any; functionAuthType?: string }

/**
 * find action method in class and return relevant node
 */
function findActionMethod(node: ts.ClassDeclaration): ts.MethodDeclaration | ts.ArrowFunction | undefined {
  const action = node.members.find((node: ts.ClassElement) => {
    if (node.name) {
      return node.name.getText() === FUNCTION_ACTION
    }
    return false
  })
  if (ts.isPropertyDeclaration(action!)) {
    return (action! as ts.PropertyDeclaration).initializer as ts.ArrowFunction
  }
  return action as ts.MethodDeclaration
}

function serializeBody(method: ts.MethodDeclaration | ts.ArrowFunction, checker: ts.TypeChecker) {
  const symbol = checker.getSymbolAtLocation(method.name)
  let data
  let error
  if (symbol) {
    const returnTypeNode = method.type!

    let dataIndex = 0 // index of ReturnedData in TFetchPromise<ReturnedData, ReturnedError>
    let errorIndex = 1 // index of ReturnedError in TFetchPromise<ReturnedData, ReturnedError>
    const typeName = checker.typeToString(checker.getTypeFromTypeNode(returnTypeNode))
    if (typeName.startsWith('Promise<TSaveStatePayload')) {
      dataIndex = 1 // index of ReturnedData in TSavePromise<State, ReturnedData, ReturnedError>
      errorIndex = 2 // index of ReturnedData in TSavePromise<State, ReturnedData, ReturnedError>
    }
    if (returnTypeNode && ts.isTypeReferenceNode(returnTypeNode)) {
      const typeData = checker.getTypeFromTypeNode(returnTypeNode.typeArguments![dataIndex])
      data = convertType(typeData, checker)
      const errorArgument = returnTypeNode.typeArguments![errorIndex]
      if (errorArgument) {
        const typeError = checker.getTypeFromTypeNode(errorArgument)
        error = convertType(typeError, checker)
      }
    }
  }
  return { data, error }
}

function resolveParameterTypeIndex(t: ts.Type, checker: ts.TypeChecker, i: number, j: number): number {
  let index = i
  if (checker.typeToString(t).startsWith('TSaveActionEvent')) {
    index = j
  }
  return index
}
/**
 *  Find the Parameter type in func and convert it to json-schema
 */
function serializeParameters(sym: ts.Symbol | undefined, node: ts.Node, checker: ts.TypeChecker) {
  if (sym) {
    const typ = checker.getTypeOfSymbolAtLocation(sym, node)
    if (typ.aliasTypeArguments) {
      const index = resolveParameterTypeIndex(typ, checker, 0, 1)
      return convertType(typ.aliasTypeArguments[index!], checker)
    }
  }
}

/**
 *  Find the auth context type and return it's name
 *  @param sym action method symbol
 *  @param node action method symbol value declaration
 *  @param checker program type checker
 */
function getFunctionAuthType(sym: ts.Symbol | undefined, node: ts.Node, checker: ts.TypeChecker): string | undefined {
  if (sym) {
    const typ = checker.getTypeOfSymbolAtLocation(sym, node)
    if (typ.aliasTypeArguments) {
      const index = resolveParameterTypeIndex(typ, checker, 1, 2)
      const authType = checker.typeToString(typ.aliasTypeArguments[index])
      if (/apiKey/.test(authType)) return 'APIKEY'
      if (/username/.test(authType) && /password/.test(authType)) return 'BASIC'
      if ((/accessToken/.test(authType) && /tokenSecret/.test(authType)) || /TOAUTH1AuthContext/.test(authType)) {
        return 'OAUTH1'
      }
      if (/accessToken/.test(authType)) return 'OAUTH2'
      if (/undefined/.test(authType)) return 'NONE'
      return authType
    }
  }
  return
}

/**
 *  Convert single func types to json schema
 *  This function tries to find the relevant type definitions, Params and `action` method resposne type
 *  and converts the types to json-schema
 *  @param functionPath absolute path to function
 *  @param options compiler options
 */
export function functionTypesToSchemaConverter(
  functionPath: string,
  options: ts.CompilerOptions = { target: ts.ScriptTarget.ES5, module: ts.ModuleKind.CommonJS }
): TOutput {
  const program = ts.createProgram([functionPath], options)
  const sourceFile = program.getSourceFile(functionPath)
  const checker = program.getTypeChecker()

  const output: TOutput = {
    requestBody: {},
    response: {}
  }

  if (sourceFile) {
    ts.forEachChild(sourceFile, visit)
  }

  return output

  function visit(node: ts.Node) {
    // Only consider exported nodes
    if (!isNodeExported(node)) {
      return
    }

    if (ts.isClassDeclaration(node) && node.name) {
      const actionMethodNode = findActionMethod(node)
      if (actionMethodNode) {
        const parameterNode = actionMethodNode.parameters[0]
        const sym = checker.getSymbolAtLocation(parameterNode.name)
        output.requestBody = serializeParameters(sym, parameterNode, checker)
        output.response = serializeBody(actionMethodNode, checker)
        output.functionAuthType = getFunctionAuthType(sym, parameterNode, checker)
      }
    }
  }

  function isNodeExported(node: ts.Node): boolean {
    return (
      (ts.getCombinedModifierFlags(node as ts.Declaration) & ts.ModifierFlags.Export) !== 0 ||
      (!!node.parent && node.parent.kind === ts.SyntaxKind.SourceFile)
    )
  }
}

/**
 *  Generate full openapi spec from functions
 *  @param functionsDir absolute path to functions directory
 *  @param functions list of func names
 *  @param integrationUuid integration unique identifier
 *  @param integrationName name of the integration
 */
export default function generator({
  functionsDir,
  functions,
  integrationUuid,
  integrationName
}: {
  functionsDir: string
  functions: string[]
  integrationUuid: string
  integrationName: string
}): string {
  const doc = topOfSpec(integrationName)
  const schemas = functions.sort().reduce((acc, func) => {
    const functionPath = path.join(functionsDir, `${func}.ts`)
    const typeSchema = functionTypesToSchemaConverter(functionPath)
    return Object.assign(
      acc,
      specPath({
        integrationUuid,
        functionName: func,
        response: { type: 'object', properties: typeSchema.response },
        requestBody: typeSchema.requestBody,
        oauth: typeSchema.functionAuthType === 'OAUTH2' || typeSchema.functionAuthType === 'OAUTH1'
      })
    )
  }, {})

  return JSON.stringify(Object.assign(doc, { paths: schemas }), null, 2)
}
