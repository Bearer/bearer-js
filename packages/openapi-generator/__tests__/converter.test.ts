import fs from 'fs'
import path from 'path'
import ts from 'typescript'

import { functionTypesToSchemaConverter as converter } from '../src/index'
const FUNCTIONS_DIR = path.join(__dirname, '__fixtures__', 'integration', 'functions')

const functions = fs.readdirSync(FUNCTIONS_DIR).map(func => [func])

describe('#functionTypeToSchemaConverter', () => {
  test.each(functions)('converts types to schemas for %s', file => {
    expect(
      converter(path.join(FUNCTIONS_DIR, file), {
        target: ts.ScriptTarget.ES5,
        module: ts.ModuleKind.CommonJS
      })
    ).toMatchSnapshot()
  })
})
