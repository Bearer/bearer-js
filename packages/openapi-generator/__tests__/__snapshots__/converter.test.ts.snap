// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#functionTypeToSchemaConverter converts types to schemas for fetchDataApiKey.ts 1`] = `
Object {
  "functionAuthType": "APIKEY",
  "requestBody": Object {
    "properties": Object {
      "id": Object {
        "type": "string",
      },
      "name": Object {
        "type": "string",
      },
      "query": Object {
        "type": "string",
      },
    },
    "type": "object",
  },
  "response": Object {
    "data": Object {
      "items": Object {
        "properties": Object {
          "comments": Object {
            "properties": Object {
              "totalCount": Object {
                "type": "number",
              },
              "visible": Object {
                "type": "boolean",
              },
            },
            "type": "object",
          },
          "id": Object {
            "type": "string",
          },
          "title": Object {
            "type": "string",
          },
        },
        "type": "object",
      },
      "type": "array",
    },
    "error": undefined,
  },
}
`;

exports[`#functionTypeToSchemaConverter converts types to schemas for fetchDataBasic.ts 1`] = `
Object {
  "functionAuthType": "BASIC",
  "requestBody": Object {
    "properties": Object {
      "id": Object {
        "type": "string",
      },
      "name": Object {
        "type": "string",
      },
      "query": Object {
        "type": "string",
      },
    },
    "type": "object",
  },
  "response": Object {
    "data": Object {
      "items": Object {
        "properties": Object {
          "comments": Object {
            "properties": Object {
              "totalCount": Object {
                "type": "number",
              },
              "visible": Object {
                "type": "boolean",
              },
            },
            "type": "object",
          },
          "id": Object {
            "type": "string",
          },
          "title": Object {
            "type": "string",
          },
        },
        "type": "object",
      },
      "type": "array",
    },
    "error": undefined,
  },
}
`;

exports[`#functionTypeToSchemaConverter converts types to schemas for fetchDataInlineType.ts 1`] = `
Object {
  "functionAuthType": "OAUTH2",
  "requestBody": Object {
    "properties": Object {
      "pullRequests": Object {
        "items": Object {
          "type": "string",
        },
        "type": "array",
      },
    },
    "type": "object",
  },
  "response": Object {
    "data": Object {
      "items": Object {
        "properties": Object {
          "comments": Object {
            "properties": Object {
              "totalCount": Object {
                "type": "number",
              },
              "visible": Object {
                "type": "boolean",
              },
            },
            "type": "object",
          },
          "id": Object {
            "type": "string",
          },
          "title": Object {
            "type": "string",
          },
        },
        "type": "object",
      },
      "type": "array",
    },
    "error": undefined,
  },
}
`;

exports[`#functionTypeToSchemaConverter converts types to schemas for fetchDataNone.ts 1`] = `
Object {
  "functionAuthType": "NONE",
  "requestBody": Object {
    "properties": Object {
      "id": Object {
        "type": "string",
      },
      "name": Object {
        "type": "string",
      },
      "query": Object {
        "type": "string",
      },
    },
    "type": "object",
  },
  "response": Object {
    "data": Object {
      "items": Object {
        "properties": Object {
          "comments": Object {
            "properties": Object {
              "totalCount": Object {
                "type": "number",
              },
              "visible": Object {
                "type": "boolean",
              },
            },
            "type": "object",
          },
          "id": Object {
            "type": "string",
          },
          "title": Object {
            "type": "string",
          },
        },
        "type": "object",
      },
      "type": "array",
    },
    "error": undefined,
  },
}
`;

exports[`#functionTypeToSchemaConverter converts types to schemas for fetchDataOauth2.ts 1`] = `
Object {
  "functionAuthType": "OAUTH2",
  "requestBody": Object {
    "properties": Object {
      "id": Object {
        "type": "string",
      },
      "name": Object {
        "type": "string",
      },
      "query": Object {
        "type": "string",
      },
    },
    "type": "object",
  },
  "response": Object {
    "data": Object {
      "items": Object {
        "properties": Object {
          "comments": Object {
            "properties": Object {
              "totalCount": Object {
                "type": "number",
              },
              "visible": Object {
                "type": "boolean",
              },
            },
            "type": "object",
          },
          "id": Object {
            "type": "string",
          },
          "title": Object {
            "type": "string",
          },
        },
        "type": "object",
      },
      "type": "array",
    },
    "error": undefined,
  },
}
`;

exports[`#functionTypeToSchemaConverter converts types to schemas for fetchDataRefType.ts 1`] = `
Object {
  "functionAuthType": "OAUTH2",
  "requestBody": Object {
    "properties": Object {
      "reference": Object {
        "type": "string",
      },
    },
    "type": "object",
  },
  "response": Object {
    "data": Object {
      "items": Object {
        "properties": Object {
          "id": Object {
            "type": "string",
          },
          "name": Object {
            "type": "string",
          },
          "totalCount": Object {
            "type": "number",
          },
        },
        "type": "object",
      },
      "type": "array",
    },
    "error": undefined,
  },
}
`;

exports[`#functionTypeToSchemaConverter converts types to schemas for fetchDataUndefined.ts 1`] = `
Object {
  "functionAuthType": undefined,
  "requestBody": undefined,
  "response": Object {
    "data": undefined,
    "error": undefined,
  },
}
`;

exports[`#functionTypeToSchemaConverter converts types to schemas for saveState.ts 1`] = `
Object {
  "functionAuthType": "OAUTH1",
  "requestBody": Object {
    "properties": Object {
      "pullRequests": Object {
        "items": Object {
          "properties": Object {
            "comments": Object {
              "properties": Object {
                "totalCount": Object {
                  "type": "number",
                },
                "visible": Object {
                  "type": "boolean",
                },
              },
              "type": "object",
            },
            "id": Object {
              "type": "string",
            },
            "title": Object {
              "type": "string",
            },
          },
          "type": "object",
        },
        "type": "array",
      },
    },
    "type": "object",
  },
  "response": Object {
    "data": Object {
      "items": Object {
        "properties": Object {
          "comments": Object {
            "properties": Object {
              "totalCount": Object {
                "type": "number",
              },
              "visible": Object {
                "type": "boolean",
              },
            },
            "type": "object",
          },
          "id": Object {
            "type": "string",
          },
          "title": Object {
            "type": "string",
          },
        },
        "type": "object",
      },
      "type": "array",
    },
    "error": undefined,
  },
}
`;
